## Slicing e Indexing em Pandas

Esta análise foi sugerida pela certificação em Ciência de Dados oferecida pela IBM.
O conjunto de dados contém dados anuais sobre os fluxos de imigrantes internacionais registrados pelos países de destino. Os dados apresentam entradas e saídas de acordo com o local de nascimento, nacionalidade ou local de residência anterior / próxima, tanto para estrangeiros como para nacionais. A versão atual apresenta dados relativos a 45 países.

Nesta análise, vamos nos concentrar nos dados de imigração canadense.

Os dados podem ser baixados [aqui](http://www.un.org/en/development/desa/population/migration/data/empirical2/migrationflows.shtml?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ).





In [7]:
#importando as bibliotecas para análise dos dados
#carregando o arquivo com os dados

import numpy as np 
import pandas as pd 

dfCanada = pd.read_excel('Canada.xlsx', sheet_name='Canada by Citizenship', skiprows=range(20), skipfooter=2)
dfCanada.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)

### Selecionar a coluna

** Existem duas maneiras de filtrar em um nome de coluna: **

Método 1: rápido e fácil, mas só funciona se o nome da coluna NÃO tiver espaços ou caracteres especiais.

```python
     df.column_name
         (retorna a série)
```

Método 2: mais robusto e pode filtrar em várias colunas.

```python
     df ['coluna']
         (retorna a série)
```

```python
     df [['coluna 1', 'coluna 2']]
         (retorna dataframe)
```

* * *

Exemplo: vamos tentar filtrar a lista de países ('Country').

In [8]:
dfCanada.Country

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Vamos tentar filtrar a lista de países ('Country') e os dados para os anos: 1980 - 1985.

In [10]:
dfCanada[['Country', 1980, 1981, 1982, 1983, 1984, 1985]]

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Selecionar a linha

Existem três maneiras principais de selecionar linhas:

```python
     df.loc [rótulo]
         #filtros pelos rótulos do índice / coluna
     df.iloc [index]
         #filtros pelas posições do índice / coluna
```

Antes de prosseguirmos, observe que o índice padrão do conjunto de dados é um intervalo numérico de 0 a 194. Isso torna muito difícil fazer uma consulta por um país específico. Por exemplo, para pesquisar dados sobre o Japão, precisamos saber o valor do índice correspondente.

Isso pode ser corrigido facilmente definindo a coluna 'País' como o índice usando o método set_index ().

In [11]:
# O oposto de set é reset. Então, para redefinir o índice, podemos usar df_can.reset_index()

dfCanada.set_index('Country', inplace=True)
dfCanada.head()

Type    Coverage  AREA Continent   REG           Region  \
Country                                                                         
Afghanistan     Immigrants  Foreigners   935      Asia  5501    Southern Asia   
Albania         Immigrants  Foreigners   908    Europe   925  Southern Europe   
Algeria         Immigrants  Foreigners   903    Africa   912  Northern Africa   
American Samoa  Immigrants  Foreigners   909   Oceania   957        Polynesia   
Andorra         Immigrants  Foreigners   908    Europe   925  Southern Europe   

                DEV             DevName  1980  1981  ...  2004  2005  2006  \
Country                                              ...                     
Afghanistan     902  Developing regions    16    39  ...  2978  3436  3009   
Albania         901   Developed regions     1     0  ...  1450  1223   856   
Algeria         902  Developing regions    80    67  ...  3616  3626  4807   
American Samoa  902  Developing regions     0     1  ...     0     0     1   
Andorra         901   Developed regions     0     0  ...     0     0     1   

                2007  2008  2009  2010  2011  2012  2013  
Country                                                   
Afghanistan     2652  2111  1746  1758  2203  2635  2004  
Albania          702   560   716   561   539   620   603  
Algeria         3623  4005  5393  4752  4325  3774  4331  
American Samoa     0     0     0     0     0     0     0  
Andorra            1     0     0     0     0     1     1  

[5 rows x 42 columns]

In [13]:
# opcional: para remover o nome do índice

dfCanada.index.name = None
dfCanada.head()

Type    Coverage  AREA Continent   REG           Region  \
Afghanistan     Immigrants  Foreigners   935      Asia  5501    Southern Asia   
Albania         Immigrants  Foreigners   908    Europe   925  Southern Europe   
Algeria         Immigrants  Foreigners   903    Africa   912  Northern Africa   
American Samoa  Immigrants  Foreigners   909   Oceania   957        Polynesia   
Andorra         Immigrants  Foreigners   908    Europe   925  Southern Europe   

                DEV             DevName  1980  1981  ...  2004  2005  2006  \
Afghanistan     902  Developing regions    16    39  ...  2978  3436  3009   
Albania         901   Developed regions     1     0  ...  1450  1223   856   
Algeria         902  Developing regions    80    67  ...  3616  3626  4807   
American Samoa  902  Developing regions     0     1  ...     0     0     1   
Andorra         901   Developed regions     0     0  ...     0     0     1   

                2007  2008  2009  2010  2011  2012  2013  
Afghanistan     2652  2111  1746  1758  2203  2635  2004  
Albania          702   560   716   561   539   620   603  
Algeria         3623  4005  5393  4752  4325  3774  4331  
American Samoa     0     0     0     0     0     0     0  
Andorra            1     0     0     0     0     1     1  

[5 rows x 42 columns]

Exemplo: vamos ver o número de imigrantes do Japão (linha 87) para os seguintes cenários:

```
1. Os dados da linha completa (todas as colunas)
2. Para o ano de 2013
3. Durante os anos 1980 a 1985
```

In [15]:
# linha completa
print(dfCanada.loc['Japan'])

Type                Immigrants
Coverage            Foreigners
AREA                       935
Continent                 Asia
REG                        906
Region            Eastern Asia
DEV                        901
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004    

In [16]:
# ano de 2013
print(dfCanada.loc['Japan', 2013])

982


In [17]:
# anos de 1980 a 1985
print(dfCanada.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]])

1980    701
1981    756
1982    598
1983    309
1984    246
1984    246
Name: Japan, dtype: object


### Filtrar com base em um critério

Para filtrar o dataframe com base em uma condição, simplesmente passamos a condição como um vetor booleano.

Por exemplo, vamos filtrar o dataframe para mostrar os dados em países asiáticos (AreaName = Asia).

In [19]:
# criar o vetor booleano com a condição

condition = dfCanada['Continent'] == 'Asia'
print(condition)

Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [20]:
# enviar a condição para o dataframe
dfCanada[condition]

Type    Coverage  AREA  \
Afghanistan                                     Immigrants  Foreigners   935   
Armenia                                         Immigrants  Foreigners   935   
Azerbaijan                                      Immigrants  Foreigners   935   
Bahrain                                         Immigrants  Foreigners   935   
Bangladesh                                      Immigrants  Foreigners   935   
Bhutan                                          Immigrants  Foreigners   935   
Brunei Darussalam                               Immigrants  Foreigners   935   
Cambodia                                        Immigrants  Foreigners   935   
China                                           Immigrants  Foreigners   935   
China, Hong Kong Special Administrative Region  Immigrants  Foreigners   935   
China, Macao Special Administrative Region      Immigrants  Foreigners   935   
Cyprus                                          Immigrants  Foreigners   935   
Democratic People's Republic of Korea           Immigrants  Foreigners   935   
Georgia                                         Immigrants  Foreigners   935   
India                                           Immigrants  Foreigners   935   
Indonesia                                       Immigrants  Foreigners   935   
Iran (Islamic Republic of)                      Immigrants  Foreigners   935   
Iraq                                            Immigrants  Foreigners   935   
Israel                                          Immigrants  Foreigners   935   
Japan                                           Immigrants  Foreigners   935   
Jordan                                          Immigrants  Foreigners   935   
Kazakhstan                                      Immigrants  Foreigners   935   
Kuwait                                          Immigrants  Foreigners   935   
Kyrgyzstan                                      Immigrants  Foreigners   935   
Lao People's Democratic Republic                Immigrants  Foreigners   935   
Lebanon                                         Immigrants  Foreigners   935   
Malaysia                                        Immigrants  Foreigners   935   
Maldives                                        Immigrants  Foreigners   935   
Mongolia                                        Immigrants  Foreigners   935   
Myanmar                                         Immigrants  Foreigners   935   
Nepal                                           Immigrants  Foreigners   935   
Oman                                            Immigrants  Foreigners   935   
Pakistan                                        Immigrants  Foreigners   935   
Philippines                                     Immigrants  Foreigners   935   
Qatar                                           Immigrants  Foreigners   935   
Republic of Korea                               Immigrants  Foreigners   935   
Saudi Arabia                                    Immigrants  Foreigners   935   
Singapore                                       Immigrants  Foreigners   935   
Sri Lanka                                       Immigrants  Foreigners   935   
State of Palestine                              Immigrants  Foreigners   935   
Syrian Arab Republic                            Immigrants  Foreigners   935   
Tajikistan                                      Immigrants  Foreigners   935   
Thailand                                        Immigrants  Foreigners   935   
Turkey                                          Immigrants  Foreigners   935   
Turkmenistan                                    Immigrants  Foreigners   935   
United Arab Emirates                            Immigrants  Foreigners   935   
Uzbekistan                                      Immigrants  Foreigners   935   
Viet Nam                                        Immigrants  Foreigners   935   
Yemen                                           Immigrants  Foreigners   935   

                                               Contin

In [21]:
# podemos passar vários critérios na mesma linha.
# vamos filtrar por AreaNAme = Asia e RegName = Southern Asia

dfCanada[(dfCanada['Continent']=='Asia') & (dfCanada['Region']=='Southern Asia')]

# nota: ao usar os operadores 'e' e 'ou', o pandas requer o uso de '&' e '|' em vez de 'e' e 'ou'
# não se esqueça de colocar as duas condições entre parênteses

Type    Coverage  AREA Continent   REG  \
Afghanistan                 Immigrants  Foreigners   935      Asia  5501   
Bangladesh                  Immigrants  Foreigners   935      Asia  5501   
Bhutan                      Immigrants  Foreigners   935      Asia  5501   
India                       Immigrants  Foreigners   935      Asia  5501   
Iran (Islamic Republic of)  Immigrants  Foreigners   935      Asia  5501   
Maldives                    Immigrants  Foreigners   935      Asia  5501   
Nepal                       Immigrants  Foreigners   935      Asia  5501   
Pakistan                    Immigrants  Foreigners   935      Asia  5501   
Sri Lanka                   Immigrants  Foreigners   935      Asia  5501   

                                   Region  DEV             DevName  1980  \
Afghanistan                 Southern Asia  902  Developing regions    16   
Bangladesh                  Southern Asia  902  Developing regions    83   
Bhutan                      Southern Asia  902  Developing regions     0   
India                       Southern Asia  902  Developing regions  8880   
Iran (Islamic Republic of)  Southern Asia  902  Developing regions  1172   
Maldives                    Southern Asia  902  Developing regions     0   
Nepal                       Southern Asia  902  Developing regions     1   
Pakistan                    Southern Asia  902  Developing regions   978   
Sri Lanka                   Southern Asia  902  Developing regions   185   

                            1981  ...   2004   2005   2006   2007   2008  \
Afghanistan                   39  ...   2978   3436   3009   2652   2111   
Bangladesh                    84  ...   2660   4171   4014   2897   2939   
Bhutan                         0  ...      1      5     10      7     36   
India                       8670  ...  28235  36210  33848  28742  28261   
Iran (Islamic Republic of)  1429  ...   6348   5837   7480   6974   6475   
Maldives                       0  ...      1      0      0      2      1   
Nepal                          1  ...    404    607    540    511    581   
Pakistan                     972  ...  13399  14314  13127  10124   8994   
Sri Lanka                    371  ...   4495   4930   4714   4123   4756   

                             2009   2010   2011   2012   2013  
Afghanistan                  1746   1758   2203   2635   2004  
Bangladesh                   2104   4721   2694   2640   3789  
Bhutan                        865   1464   1879   1075    487  
India                       29456  34235  27509  30933  33087  
Iran (Islamic Republic of)   6580   7477   7479   7534  11291  
Maldives                        7      4      3      1      1  
Nepal                         561   1392   1129   1185   1308  
Pakistan                     7217   6811   7468  11227  12603  
Sri Lanka                    4547   4422   3309   3338   2394  

[9 rows x 42 columns]